File: global localization/recording/python/display/generic_display.py
from display.gl_viewer import GLViewer
from exporter.KMLExporter import *
import time


class GenericDisplay:
    def __init__(self):
        pass

    def __del__(self):
        closeAllKMLFiles()

    def init(self, camera_model):
        self.glviewer = GLViewer()
        self.glviewer.init(camera_model)
        # Replace this part with the appropriate connection to your IoT system

    def updatePoseData(self, zed_rt, str_t, str_r, state):
        self.glviewer.updateData(zed_rt, str_t, str_r, state)

    def isAvailable(self):
        return self.glviewer.is_available()

    def updateRawGeoPoseData(self, geo_data):
        try:
            # Replace this part with the appropriate sending of data to your IoT system
            latitude, longitude, _ = geo_data.get_coordinates(False)
            f = open('../../map server/raw_data.txt', 'w')
            f.write("{},{},{}".format(latitude, longitude, geo_data.ts.get_milliseconds()))

        except ImportError:
            print("An exception was raised: the raw geo-pose data was not sent.")

    def updateGeoPoseData(self, geo_pose, current_timestamp):
        try:
            # Replace this part with the appropriate sending of data to your IoT system
            f = open('../../map server/data.txt', 'w')
            f.write("{},{},{}"
                    .format(geo_pose.latlng_coordinates.get_latitude(False),
                            geo_pose.latlng_coordinates.get_longitude(False),
                            current_timestamp.get_milliseconds()))

        except ImportError:
            print("An exception was raised: the geo-pose data was not sent.")


if __name__ == "__main__":
    generic_display = GenericDisplay()
    generic_display.init(0, [])

    try:
        while True:
            # Your logic here...
            pass
    except KeyboardInterrupt:
        pass

File: global localization/recording/cpp/src/display/GenericDisplay.cpp
#include "display/GenericDisplay.h"
#include "exporter/KMLExporter.h"


GenericDisplay::GenericDisplay()
{
}

GenericDisplay::~GenericDisplay()
{
    closeAllKMLWriter();
}

void GenericDisplay::init(int argc, char **argv)
{
    opengl_viewer.init(argc, argv);
}

void GenericDisplay::updatePoseData(sl::Transform zed_rt, sl::FusedPositionalTrackingStatus state)
{
    opengl_viewer.updateData(zed_rt, state);
}

bool GenericDisplay::isAvailable(){
    return opengl_viewer.isAvailable();
}

void GenericDisplay::updateRawGeoPoseData(sl::GNSSData geo_data)
{
    double latitude, longitude, altitude;
    geo_data.getCoordinates(latitude, longitude, altitude, false);

    // Make the pose available for the Live Server
    ofstream data;
    data.open ("../../../map server/raw_data.txt");
    data << std::fixed << std::setprecision(17);
    data << latitude;
    data << ",";
    data << longitude;
    data << ",";
    data << geo_data.ts.getMilliseconds();
    data << "\n";
    data.close();
}

void GenericDisplay::updateGeoPoseData(sl::GeoPose geo_pose, sl::Timestamp current_timestamp)
{
    // Make the pose available for the Live Server
    ofstream data;
    data.open ("../../../map server/data.txt");
    data << std::fixed << std::setprecision(17);
    data << geo_pose.latlng_coordinates.getLatitude(false);
    data << ",";
    data << geo_pose.latlng_coordinates.getLongitude(false);
    data << ",";
    data << current_timestamp.getMilliseconds();
    data << "\n";
    data.close();

    // Save the pose in a .kml file
    saveKMLData("fused_position.kml", geo_pose);
}
